pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -158,6 +158,18 @@ func newDefaultQueueSort() framework.LessFunc {
 	return sort.Less
 }
 
+// mockSchedCycleWithFailure mocks a scheduling cycle to schedule a Pod with failure:
+// - put the given Pod to q.processingPods
+// - call q.AddUnschedulableIfNotPresent()
+// - remove the given Pod from q.processingPods
+// NOTE: if the unit test uses Pop() to pop a Pod, you can directly call AddUnschedulableIfNotPresent()
+// as Pod() already put the popped Pod to q.processingPods.
+func mockSchedCycleWithFailure(q *PriorityQueue, pod *v1.Pod) {
+	q.processingPods[pod.UID] = struct{}{}
+	q.AddUnschedulableIfNotPresent(newQueuedPodInfoNoTimestamp(pod), q.SchedulingCycle())
+	delete(q.processingPods, pod.UID)
```

Comment:
I feel like this delete should be part of `AddUnschedulableIfNotPresent`. Basically same rationale as checking for processingPods not empty.

Now, you might ask what happens if the pod gets scheduled. And actually, it looks like our event handlers handle that already. 2 functions get called when a pod passes from unscheduled to scheduled: `sched.addPodToCache` and `sched.deletePodFromSchedulingQueue`.

https://github.com/kubernetes/kubernetes/blob/b48f5af2602bf097c11c6a252bc7f82a7c096206/pkg/scheduler/eventhandlers.go#L353
https://github.com/kubernetes/kubernetes/blob/b48f5af2602bf097c11c6a252bc7f82a7c096206/staging/src/k8s.io/client-go/tools/cache/controller.go#L264

Reply:
Technically we can do that way. But I'm thinking about whether it outweighs the current fix. 

Current fix simply adds a processing Pod into the `processingPods` map at the beginning of a cycle (in `Pop()`), and ensure to remove it at the end of schedulingCycle or bindingCycle, which can be easily understood. However, if we turn to removing it in `AddUnschedulableIfNotPresent`, it causes extra places to take care - the places that end a cycle but don't call `AddUnschedulableIfNotPresent` -  which needs more mental efforts, and _may be_ error-prone along with the codebase development. WDYT?

Reply:
I wasn't too concerned about `q.processingPods[pod.UID] = struct{}{}` as much as I was for `delete(q.processingPods, pod.UID)`.
This is fine for now.